payment_gateway_adyen/services/payment_service.py

```diff
@@ -0,0 +1,244 @@
+        except AdyenAPIInvalidAmount as e:
+            self._raise_error_message('invalid_amount')
+        except AdyenAPIInvalidFormat as e:
+            self._raise_error_message('validation')
+        except (AdyenAPIResponseError, AdyenAPIValidationError) as e:
+            if e.error_code:
+                code = e.error_code
+            else:
+                match = re.search('errorCode: ([0-9][0-9][0-9][0-9]?)',
+                                  e.message)
+                if match:
+                    code = match.group(1).strip()
+                else:
+                    code = 'undef'
+            self._raise_error_message(code)
+        except Exception as e:
+            self._raise_error_message('undef')
+
+    def process_return(self, **params):
+        payload = {}
+        payload["browserInfo"] = {
+            "userAgent": params.get(
+                "userAgent",
+                "Mozilla/5.0 (X11; U; Linux i686; en-US; rv:\
+                1.9) Gecko/2008052912 Firefox/3.0"),
+            "acceptHeader": params.get(
+                "acceptHeader",
+                "text/html,application/xhtml+xml,\
+                application/xml;q=0.9,*/*;q=0.8")
+        }
+        payload["md"] = params['MD']
+        payload["paResponse"] = params['PaRes']
+        result = self._http_adyen_request('authorise3d', payload)
+        vals = {
+            'state': MAP_SOURCE_STATE[result.message['resultCode']],
+            'data': json.dumps(result.message),
+            }
+
+        transaction = self.env['gateway.transaction'].search([
+            ('external_id', '=', result.message['pspReference']),
+            ('payment_mode_id.provider', '=', 'adyen'),
+            ])
+        if transaction:
+            transaction.write(vals)
+        else:
+            raise UserError(
+                _('The transaction %s do not exist in Odoo') % result.message[
+                    'pspReference'])
+
+    def _get_formatted_amount(self, amount=None):
+        if amount is None:
+            amount = self.collection._get_amount_to_capture()
+        if self.collection.currency_id.name in ZERO_DECIMAL_CURRENCIES:
+            return int(amount)
+        elif self.collection.currency_id.name in THREE_DECIMAL_CURRENCIES:
+            return int(float_round(amount * 1000, 0))
+        else:
+            return int(float_round(amount * 100, 0))
+
+    def _prepare_charge(self, source=None, **kwargs):
+        transaction = self.collection
+        description = "|".join([
+            transaction.name,
+            transaction.partner_id.email,
+            str(transaction.id)])
+        vals = {
+            'amount': {
+                "value": self._get_formatted_amount(),
+                "currency": transaction.currency_id.name},
+            'reference': description,
+            'additionalData': {"executeThreeD": "true"}  # optional may be?
+            }
+        if kwargs.get('encrypted_card'):
```

Comment:
You can always consider that the encrypted card is passed. The check will be done before in shopinvader module.

Regarding the 3D, you should pass the browserinfo here in order to apply dynamic 3D https://docs.adyen.com/developers/risk-management/3d-secure#authorisethe3dsecurepayment

What you can do is to check if the key "browserinfo" exist in kwarg (shopinvader will always add the key) if yes we apply dynamic 3DS by passing the browser info if not we apply set to false the 3DS.

Reply:
ok, so if we use a browserinfo dict here, to be consistent, I assume I should also change the process_return method to use browserinfo too instead of the two inner key/values ```userAgent``` and ```acceptHeader``` that I am using at the moment. Right?